File: leet16/leet16.py
# https://leetcode.com/problems/3sum-closest/

from typing import List

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        close_sum = float("inf")
        n = len(nums)
        for i in range(n - 2):
            left, right = i + 1, n - 1
            while left < right:
                total_sum = nums[i] + nums[left] + nums[right]
                if abs(total_sum - target) < abs(close_sum - target):
                    close_sum = total_sum
                if total_sum == target:
                    return total_sum
                elif total_sum > target:
                    right -= 1
                else:
                    left += 1
        return close_sum

File: leet16/leet16.cpp
// https://leetcode.com/problems/3sum-closest/description/

#include <vector>
#include <algorithm>
using namespace std;

class Solution
{
public:
    int threeSumClosest(vector<int> &nums, int target)
    {
        sort(nums.begin(), nums.end());
        int left, right, close_sum = nums[0] + nums[1] + nums[2], total_sum;
        int n = nums.size();
        for (int i = 0; i < n - 2; i++)
        {
            left = i + 1;
            right = n - 1;
            while (left < right)
            {
                total_sum = nums[i] + nums[left] + nums[right];
                if (total_sum == target)
                {
                    return total_sum;
                }
                else if (total_sum > target)
                {
                    right--;
                }
                else
                {
                    left++;
                }
                if (abs(total_sum - target) < abs(close_sum - target))
                {
                    close_sum = total_sum;
                }
            }
        }
        return close_sum;
    }
};